WooCommerce Shipping order page banner prompt

Ferdev opened the pull request:
Implements a banner that is displayed in the WooCommerce order page, whenever *all* these conditions are met:

- WooCommerce Shipping banner NOT dismissed
- Jetpack >= 4.4 installed and active
- Jetpack Connected
- No incompatible extensions installed:
   - Shipstation not installed.
   - UPS not Installed.
   - Fedex not installed.
   - ShippingEasy not installed.
- Order contains physical products which need to be shipped ([we should check that the order status is not set to complete](https://a8c.slack.com/archives/CKNC7KACA/p1584395171186200)).
- Store is located in US.
- Store currency is set to USD.
- `WCS plugin not installed` OR` WCS is installed *AND* ToS have NOT been accepted *AND* WCS version is 1.22.5 or greater`.
    - The 1.22.5 or greater requirement is so we can launch the shipping modal from the banner.

### Screenshots

https://www.dropbox.com/s/m8dpajr0w1lyo8v/shot-2020-03-20-15-31-32.png?dl=0
https://www.dropbox.com/s/7mkrlxk5tnva29a/shot-2020-03-20-15-31-02.png?dl=0

### Detailed test instructions:

1. Setup your environment considering all the conditions described in this PR's summary, or to make it easy:
   1. Start with a fresh WooCommerce site (WooCommerce Services not installed)
   2. Install and Connect Jetpack
   3. Install `shipping-bannner` branch of wc-admin
2. Create an order with a shippable product.
3. Go to that order's page (**Note: the A/B testing logic only shows the prompt to 25% of sites. Set to the option `woocommerce_shipping_prompt_ab` to `b` to enable it for your test site**)
4. Press the create shipping label button.
5. The WooCommerce Services create shipping label modal should pop up, and a new label should be able to be purchased.


### Changelog Note:

superdav42 committed:
Initial work on onboarding print shipping banner on edit order page.

superdav42 committed:
Add all external scripts as requirements to fix rendering

superdav42 committed:
Only depend on moment as it is external is webpack.config.js and wordpress/components is not

superdav42 committed:
Update banner text

superdav42 committed:
Implement more parts of banner

c-shultz committed:
Shipping banner: Move dismiss modal into separate component. (#3837)

superdav42 committed:
Update to match layout

superdav42 committed:
Update with new design

superdav42 committed:
Finish mobile size designs and modal

c-shultz committed:
Shipping banner: Use wc-admin api to save banner dismissed option. (#3838)

superdav42 committed:
Remove unused items

superdav42 committed:
Merge branch 'shipping-banner' of github.com:woocommerce/woocommerce-admin into shipping-banner

Commit:
Add/issue 15 trackings (#3839)

* Add tracking for shipping banner

* Add props to tracking function since refactored

* Only track one view per banner render

* Retrieve plugin and jetpack statuses

* Use wc-api to retrieve plugin statuses

* Remove data store initiation to avoid duplication

* Use onboarding's activePlugins if not preloaded

* Remove banner-track file since it's no longer needed

* Add unit test for tracking events on shipping banner

* Add trackBannerEvent to destructure

* Add test for dismiss-modal

* Updated tests for interpolateComponents

Also cleaned up some unused codes

* Update test copy

superdav42 committed:
Remove redundance styles

c-shultz committed:
Onboarding banner: use an isset check to add activePlugins data for robustness. (#3900)

c-shultz committed:
Add shipping banner error notice (#3902)

Commit:
Install and acitvate WCS from banner (#3901)

* Load onboarding API if shipping-banner-feature is enabled

Allow one-click to install WCS in banner.

* Install and activate WCS through wc-api

* Add activation and installation props to tests

* Add test for installation and activation plugin

superdav42 committed:
Hide external link icon (#3920)

Commit:
Display error if there are errors sent back by wc-api (#3919)

Commit:
Show loading state when installing or activating WCS (#3915)

* Show loading state when installing or activating WCS

* Add installation and activation errors init props to test

superdav42 committed:
Reduce asset sizes for shipping banner (#3912)

* Use Dashicons everywhere to save space only include styles needed
closes https://github.com/Automattic/woocommerce-shipping-issues/issues/30

superdav42 committed:
Use appropriate line height to fix display issue in Firefox (#3932)

Ferdev committed:
New/shipping banner display logic 13 (#3852)

* Create class to manage shipping banner display

c-shultz committed:
Add a/b test to enable prompt for 25% of eligible sites. (#3933)

Commit:
Fix/issue 29 tracking (#3934)

* Convert events to impressions and click

* Update test to the new events

Ferdev committed:
Launch WCS label modal from wc-admin (#3937)

* Add methods to open WCS modal from shipping banner

* Implement test for the shipping banner component

* Add a data store for the shipping banner component

* Add test to shipping banner component

* Update Shipping Banner WC Services API endpoints

* Fix broken test for Shipping Banner

* Remove unneded call to loadWcsAssets

* Remove comments

* Fix error loading WC Services assets

* Avoid loading WC Services assets more than once

* Add global JS variables required by WC Services

* Load ToS and Assets requests sequentially.

* Remove unneeded comment

Commit:
Hide the shipping banner when dismiss is clicked (#3944)

Commit:
Update copy to WooCommerce Shipping (#3947)

superdav42 committed:
Shipping banner load in the proper place (#3950)

* Load the shipping banner and tracking sidebar in the appropriate places

* Hide wc-admin shipping banner

Commit:
Keep "Create shipping label" busy until asset is done loading (#3951)

c-shultz committed:
Shipping banner: skip install if already installed and show err… (#3952)

* Hide WCS 'Connect' banner when displaying onboarding prompt on order page.

* Move to direct api fetch instead of data store to simplify async handling for now.

* Show error for WCS set up issues.

Ferdev committed:
Fix broken JS test for the WC Services banner.

Ferdev committed:
Fix broken phpunit tests

Ferdev committed:
Fix broken JS tests for the WC Services banner.

superdav42 committed:
Merge branch 'master' of github.com:woocommerce/woocommerce-admin into shipping-banner

timmyc commented on client/wp-admin-scripts/print-shipping-label-banner/dismiss-modal/index.js:
```diff
@@ -0,0 +1,81 @@
+
+	hideBanner = () => {
+		document.getElementById(
+			'woocommerce-admin-print-label'
```
Any chance this element might not be in the DOM, should we check just in case?

superdav42 replied:
No, not really. If it's not in the dom then we could never get here.

timmyc commented on client/wp-admin-scripts/print-shipping-label-banner/dismiss-modal/index.js:
```diff
@@ -0,0 +1,81 @@
+
+		return (
+			<Modal
+				title="Are you sure?"
```
Guessing this needs to be translated?

superdav42 replied:
Yes

timmyc commented on client/wp-admin-scripts/print-shipping-label-banner/shipping-banner/index.js:
```diff
@@ -0,0 +1,448 @@
+			activatePlugins( [ wcsPluginSlug ] );
+		}
+		if ( this.justActivatedWcs( prevProps ) ) {
+			// TODO: Add success notice after installation #32
```
Has this been done?

superdav42 replied:
Yes

timmyc commented on client/wp-admin-scripts/print-shipping-label-banner/shipping-banner/index.js:
```diff
@@ -0,0 +1,448 @@
+		);
+	}
+}
+
```
Would be nice to have a `propTypes` declaration somewhere for this component. Guessing it would help enforce a few props as required that are anticipated.

superdav42 replied:
I'll add one

timmyc commented on client/wp-admin-scripts/print-shipping-label-banner/shipping-banner/index.js:
```diff
@@ -0,0 +1,448 @@
+	}
+
+	hasActivationError = () => {
+		return Boolean( this.props.activationErrors.length );
```
Do we need to verify this prop is present ( and an array ) - or perhaps we add some required propType declarations to help protect.

superdav42 replied:
Proptypes will take care of it

timmyc commented on client/wp-admin-scripts/print-shipping-label-banner/shipping-banner/index.js:
```diff
@@ -0,0 +1,448 @@
+	};
+
+	hasInstallationError = () => {
+		return Boolean( this.props.installationErrors.length );
```
Likewise here.

superdav42 replied:
Proptypes will take care of it

timmyc commented on client/wp-admin-scripts/print-shipping-label-banner/shipping-banner/index.js:
```diff
@@ -0,0 +1,448 @@
+			.insertAdjacentHTML( 'afterend', shipmentTrackingHtml );
+
+		// Need to refresh so the new metaboxes are sortable.
+		window.jQuery( '#normal-sortables' ).sortable( 'refresh' );
```
Perhaps my optimism here is too great, but should we wrap these lines in an `if ( window.jQuery )` for the maybe someday when it will no longer be present?

superdav42 replied:
Can't hurt

timmyc commented on client/wp-admin-scripts/print-shipping-label-banner/shipping-banner/index.js:
```diff
@@ -0,0 +1,448 @@
+	}
+
+	componentDidMount() {
+		this.trackImpression();
```
Since the `render()` returns `null` if `! showShippingBanner`, we will still be tracking an impression even if the banner isn't rendered. Not sure if that is the intent of this track, but based upon the name... thinking `this.trackImpression()` should only be called if `showShippingBanner` is true?

superdav42 replied:
showShippingBanner defaults to true so it'll always show at first unless it's dismissed. We have the PHP code hiding it completely so it shouldn't track the impression unless it's actually shown. However I think it's a good idea to show this check incase we ever change how it works.

timmyc commented on client/wp-admin-scripts/print-shipping-label-banner/shipping-banner/index.js:
```diff
@@ -0,0 +1,448 @@
+						alt={ __( 'Shipping ', 'woocommerce-admin' ) }
+					/>
+					<Button
+						disabled={ this.state.isShippingLabelButtonBusy }
```
Since we are already deconstructing `this.state` above, perhaps we could add in `isShippingLabelButtonBusy` and `isShippingLabelButtonBusy` as well to tidy things up here a bit?

superdav42 replied:
yes

timmyc commented on client/wp-admin-scripts/print-shipping-label-banner/shipping-banner/index.js:
```diff
@@ -0,0 +1,448 @@
+						isBusy={ this.state.isShippingLabelButtonBusy }
+						onClick={ this.createShippingLabelClicked }
+					>
+						{ __( 'Create shipping label' ) }
```
Missing text domain on this translation call.

timmyc commented on client/wp-admin-scripts/print-shipping-label-banner/shipping-banner/test/index.js:
```diff
@@ -0,0 +1,391 @@
+		);
+	} );
+
+	it( 'should record an event when user sees banner loaded', () => {
```
Related to my comment above around the impression tracking in `componentDidMount` depending on the needs of that track - i.e. if we only want it recorded if the banner is visible, maybe we can add a test to verify that here.

timmyc commented on config/development.json:
```diff
@@ -6,6 +6,7 @@
 		"onboarding": true,
-		"store-alerts": true
+		"store-alerts": true,
+	    "shipping-label-banner": true
```
Looks like the indentation here is a bit off.

timmyc commented on package-lock.json:
```diff
@@ -21196,7 +21196,7 @@
     },
     "prettier": {
-      "version": "npm:wp-prettier@1.19.1",
+      "version": "npm:prettier@1.19.1",
```
Not a big deal but typically we omit any package-lock.json changes on branches.

timmyc commented on src/Features/ShippingLabelBanner.php:
```diff
@@ -0,0 +1,188 @@
+<?php
+/**
+ * WooCommerce Onboarding Tasks
```
Little copy/paste remnants here :)

timmyc commented on src/Features/ShippingLabelBanner.php:
```diff
@@ -0,0 +1,188 @@
+	public function meta_box( $post, $args ) {
+
+		?>
+		<div id="wc-admin-shipping-banner-root" class=" woocommerce <?php echo esc_attr( 'wc-admin-shipping-banner' ); ?>" data-args="<?php echo esc_attr( wp_json_encode( $args['args'] ) ); ?>">
```
minor nit but does the extra space before woocommerce in the class list need to be there?

superdav42 replied:
no

timmyc commented on src/Features/ShippingLabelBannerDisplayRules.php:
```diff
@@ -0,0 +1,224 @@
+<?php
+/**
+ * WooCommerce Onboarding Tasks
```
Another ref to onboarding here.

timmyc commented on tests/features/class-wc-tests-shipping-label-banner-display-rules.php:
```diff
@@ -0,0 +1,306 @@
+/**
+ * Class WC_Tests_Shipping_Label_Banner_Display_Rules
+ */
+class WC_Tests_Shipping_Label_Banner_Display_Rules extends WC_Unit_Test_Case {
```
Thanks for all the test coverage in this PR, super awesome.

timmyc reviewed (commented):
Will take this for a test drive soon, but some comments on the code. Overall this is fantastic, and as mentioned already 🙌 for all the test coverage!

timmyc commented:
Oh @Ferdev / @c-shultz this branch needs a rebase against `master` too.

timmyc commented:
Did a test drive on a fresh ephemeral site. Verified a/b functionality, and that the prompt is not shown until Jetpack is connected. Verified tracking pixel was dropped when I clicked the WooCommerce Services Link.

Install/activate was quick! Saw the tracks for that too.

elizaan36 commented:
Nice work y'all. Just tested on a new Jurassic ninja site and [followed all the steps here](https://heisenmattic.wordpress.com/2020/03/20/wcship-order-page-prompt-update-ready-for-review/).

A couple things I noticed:

1. I'm wondering if we can vertically center the text on larger screen sizes. The text is too high in the prompt card. I'd love to help tweak to save you all from having to do it, but don't know if it's going to take me too much time to set up.

<img width="1433" alt="Screen Shot 2020-03-20 at 3 26 29 PM" src="https://user-images.githubusercontent.com/5121465/77204603-11516200-6ac1-11ea-8b31-532df920effa.png">

2. After I clicked "Create shipping label" it looked like the "shipping label" meta box appeared for a split second, in two different places before the modal appears.

At :04 seconds during activation
<img width="1673" alt="Screen Shot 2020-03-20 at 3 33 36 PM" src="https://user-images.githubusercontent.com/5121465/77205845-b5d4a380-6ac3-11ea-89e2-be0f6d486696.png">

At :05 seconds during activation
<img width="1674" alt="Screen Shot 2020-03-20 at 3 34 08 PM" src="https://user-images.githubusercontent.com/5121465/77205853-b8cf9400-6ac3-11ea-96b5-2552fe2c0a45.png">

superdav42 committed:
Fix building zip file in Linux

superdav42 committed:
Add proptypes and a little code cleanup

superdav42 commented:
> I'm wondering if we can vertically center the text on larger screen sizes. The text is too high in the prompt card. I'd love to help tweak to save you all from having to do it, but don't know if it's going to take me too much time to set up

I think we could do this without too much effort.

> After I clicked "Create shipping label" it looked like the "shipping label" meta box appeared for a split second, in two different places before the modal appears.

Are you suggesting we hide the banner from wc-admin before the WCS banner loads?
It might be tricky but we could try. Probably best to hide the wcs banner before it's fully loaded. That should be easy enough.

elizaan36 commented:
> I think we could do this without too much effort.

Awesome.

> Are you suggesting we hide the banner from wc-admin before the WCS banner loads?
It might be tricky but we could try. Probably best to hide the wcs banner before it's fully loaded. That should be easy enough.

Oh yeah, I should have mentioned what I expected to see.😅 I agree it would be best to hide the WCS banner until the modal opens. Would feel like a cleaner setup with less noise on the page to try and understand. It's a little disorienting to have the second banner appear, and then when the button changes from pink to blue, all before the modal opens.

Commit:
Don't change the banner text during WCS installation/activation (#3979)

timmyc commented:
@superdav42 @Ferdev is this ready for another set of 👀 yet?

Ferdev committed:
Hide the WC Shipping banner until the create shipping label pop… (#3975)

* Hide WC Shipping banner until the modal is open

* Replaces the MutationObserver in Shipping banner

Instead, it uses a subscription to the WooCommerce Services store.

Commit:
Add flex to banner to center vertical text alignment (#3976)

* Add flex to banner to center vertical text alignment

* Removed bottom padding in mobile view

* Add equal padding to top and bottom of banner

c-shultz commented:
@timmyc,

Now would be a great time to get more eyes on this.

timmyc commented on client/wp-admin-scripts/print-shipping-label-banner/shipping-banner/index.js:
```diff
@@ -0,0 +1,499 @@
+	}
+}
+
+ShippingBanner.propTypes = {
```
Thanks.

timmyc commented on src/API/Init.php:
```diff
@@ -74,7 +74,7 @@ public function rest_api_init() {
 		);
 
-		if ( Loader::is_onboarding_enabled() ) {
+		if ( Loader::is_onboarding_enabled() || Loader::is_feature_enabled( 'shipping-label-banner' ) ) {
```
Could we add a comment here that details which endpoint(s) are used by the shipping label banner just so if/when in the future we refactor this a bit, it will be easy to see why this change was made?

timmyc commented on bin/make-zip.sh:
```diff
@@ -21,9 +21,9 @@ output 2 "Creating archive... 🎁"
 ZIP_FILE=$1
 
-build_files=$(ls dist/*/*.{js,css})
+build_files=$(find dist/ \( -name '*.js' -o -name '*.css' \))
```
Curious about the changes here, was something not working as expected?

superdav42 replied:
Yes, the command was failing for me in Linux. Not all shells support the `/*/*.` glob so find should be much more portable.

timmyc replied:
Thanks for fixing that up @superdav42

timmyc reviewed (approved):
Took this for another test on a brand new AT site, and it is working quite nicely. Few minor comments in the code, but thanks for all the updates already. Would love to get another set of eyes here and 👍 prior to merge.

superdav42 committed:
Only load the controller used by the shipping banner

superdav42 committed:
Correct comments

superdav42 committed:
Merge branch 'master' of github.com:woocommerce/woocommerce-admin into shipping-banner

timmyc commented:
Looks like we have some failing tests here now.

superdav42 committed:
Mock some more methods we started using

Commit:
Position error message directly under the paragraph (#4010)

Ferdev committed:
Fix/remove address alerts 55 (#3996)

* Open the WC Services modal once initialized

Waits until the WooCommerce Services extension has loaded its initial
state properly.

* Change the way the WC Services modal is open

Instead of dispatching actions, the banner now triggers the click event
on the "create shipping label" button.

Ferdev committed:
Hide scrollbars in dismiss modal (#4008)

* Hide scrollbars in dismiss modal

* Add back css import removed by accident.

timmyc commented:
Tests are green, I'm totally cool with you all landing this anytime you are ready @c-shultz  / @Ferdev  / @superdav42

c-shultz commented:
Hey, @timmyc, we're reviewing one last PR on this regarding the banner dismissal logic: https://github.com/woocommerce/woocommerce-admin/pull/4018

We should be good to go in the next couple hours.

Commit:
Fix/shipping banner dismiss (#4018)

* Use milliseconds for dismiss timing and convert from string.

* Fix test

Co-authored-by: c-shultz <chris.shultz@automattic.com>
